### ISSUE TITLE
PHPStorm on Windows, DDEV in Docker in WSL2, Docker-Compose path wrong

### ISSUE THREAD

**Issue by ChristianAdamski** [php-perfect/ddev-intellij-plugin#152]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Are you sure that this bug is related to this DDEV Integration Plugin?

- [X] I am sure

### Describe the bug

![phpstorm](https://user-images.githubusercontent.com/8464825/216004176-c15e64c1-070b-453f-8564-dc313eeb9cab.png)

See above image. While the path to the yml is correct, something in the composer call is wrong. Paths set in Settings->Build -> Docker -> Tools have no effect and wouldnt work anyway, as these require Windows executables.

**Comment by rfay** [comment]
@ChristianAdamski and I spent quite a bit of time on this and I explained that this is a PhpStorm regression (Unable to configure docker-compose properly in docker settings). I guess I wasn't clear. 

@ChristianAdamski please simplify the recreation of this PhpStorm problem (that you can't configure docker-compose properly) and then search issues in https://youtrack.jetbrains.com/issues/WI and then if there's nothing there, file a bug about this regression.

**Comment by ChristianAdamski** [comment]
No, I'm actually sure this is an issue of this plugin. I did start from scratch and also tested PHPStorm within WSL, as well as EAP versions of PHPStorm.

I do notice, that the config path is set here:

https://github.com/php-perfect/ddev-intellij-plugin/blob/main/src/main/java/de/php_perfect/intellij/ddev/php/DdevComposeFileLoaderImpl.java

I have not yet figured out, where the docker-compose path is set.

**Comment by rfay** [comment]
Are you able to configure the docker integration successfully if the plugin is removed?

**Comment by ChristianAdamski** [comment]
No, not at all. Which brings me to the question, why PHPStorm is able to locate the file at all, if I never point to it anywhere. Hence my assumption, that the DDEV plugin is taking care of it,

**Comment by ChristianAdamski** [comment]
The "DDEV" CLI interpreter is automagically configured by the Module. As is the server settings.

**Comment by rfay** [comment]
@ChristianAdamski you're still using docker-ce inside wsl2 right, which has never been claimed to be supported by the plugin. Have you tried with Docker Desktop?

**Comment by ChristianAdamski** [comment]
Well, this is effectively a feature request. I just didn't see that as an option. I'm not saying this plugin is broken on something it advertises to do, I'm saying it _might_ be fairly easy to fix this here. I do not know though.

**Comment by rfay** [comment]
If I understand your goal, it's to run the php interpreter inside the DDEV web container. 

The failure here has nothing to do with this plugin. Here are my results of trying to configure the PHP interpreter in the traditional way when using WSL2 as the docker provider. This is with the plugin disabled.

![image](https://user-images.githubusercontent.com/112444/216111393-b165f797-4ba9-46e4-b987-7caa88796891.png)

I'm pretty sure you need to file an issue with PhpStorm. (BTW, I tested on macOS with alternate docker provider colima, and it worked there.) And yes, PhpStorm is very confused about what docker-compose it can or should use.

**Comment by rfay** [comment]
I confirm that (without plugin) Docker Desktop works fine to use the in-container php cli, as normally documented in the DDEV docs. It does *not* work with docker-ce. None of this has to do with this extension, although if PhpStorm work working correctly we might find out. 

To proceed, it's going to be necessary to demonstrate this problem in the PhpStorm queue, it won't be solved here or in the DDEV queue. 

To get them to look at it, you'll need to use a trivial `docker-compose.yaml` and show them the steps to get the failure.

**Comment by nico-loeber** [comment]
Hi @ChristianAdamski,

Thank you for reporting your issue!
As @rfay already pointed out the DDEV PHP Interpreter is just an automatically configured interpeter without any special magic.

The docker compose binary that is used can be configured in `File | Settings | Build, Execution, Deployment | Docker | Tools`, this is not in the scope of the plugin.
Please make sure to check `Use compose V2` there, as ddev is using docker compose v2 internally.

As already mentioned above, supportet system configurations are Ubuntu with docker-ce, Mac with docker or colima and Windows with Docker Desktop, both inside WSL or on the native file system.

Sadly @rfay just showed us that PHPStorm doesn't seem to support your setup (docker-ce in WSL) yet.

As soon as there is support by the IDE we can see if there is anything we can optimize in the plugin.

If there is anything I can help you with, please let me know.

Thanks @rfay for your fast reaction on this and the testing!

**Comment by rfay** [comment]
Side-note FYI @nico-loeber "compose V2" setting in Docker Desktop is misinformed/mislabeled. What they mean is "Use `docker compose` instead of `docker-compose`". And it doesn't affect what docker-compose version they use. PhpStorm just isn't paying careful attention to real things :)

**Comment by nico-loeber** [comment]
Yes I am aware of that @rfay, but using the v2 syntax will make sure the installed docker compose is v2 ready. Depending on your OS the versions delivered by your package manager may not be v2.
Also you never know which changes might come to the next IDE release. So when using docker-compose v2 it's a good idea to tell your IDE.

**Comment by rfay** [comment]
What I'm saying is that the checkbox here does not say anything about v2 syntax.

* The checkbox just makes it (apparently, according to the text) use `docker compose` (which would not work without the `compose` plugin to v2). 
* However, they let you explicitly specify the `docker-compose` binary, which we do.
* DDEV does not support any `docker-compose` v1 any more for anything. 
* DDEV does not ever use the docker-compose provided by the OS package manager any more.

**Comment by rfay** [comment]
Actually, the more I think about it:

If you check the `v2` checkbox in PhpStorm... it uses the `docker` client and issues `docker compose` which uses the `compose` plugin and thus does *not* use the configured  (and tested) DDEV `docker-compose` version. Instead, it uses the OS-supplied (or not) `compose` plugin.

**Comment by nico-loeber** [comment]
Interesting thoughts @rfay but a bit offtopic. ;)

**Comment by alexbogias** [comment]
Nothing here yet but loos like is only Jetbrains issue!

**Comment by nico-loeber** [comment]
As @shochdoerfer mentioned, the fix of the issue #187 might have impact on this as well, since there were missing environment varialbes that could cause different "binary x not found" errors under certain circumstances.
Please let me know if you still have this issue in PHPStorm 2023.2 with version 1.1.1 of the plugin @ChristianAdamski, thanks!

**Comment by rfay** [comment]
I need to open an issue with Jetbrains about the misunderstanding of "v2" :)

**Comment by rfay** [comment]
Youtrack issue about misunderstanding "Compose V2": https://youtrack.jetbrains.com/issue/WI-73632/Misundertanding-Miscommunication-of-docker-compose-v2

**Comment by alexbogias** [comment]
One year letter and after many JetBrains and WSL tickets... TADA 
![image](https://github.com/php-perfect/ddev-intellij-plugin/assets/1313467/4e937526-265f-4fb7-b34e-3c0cb6a87e2b)

**Comment by rfay** [comment]
Congrats to @nico-loeber ! 

I still don't know how docker-compose can properly be used with PhpStorm on Windows and WSL2. It always wants to use a docker-compose.exe on the WIndows side, which sometimes even works, but it's just wrong. I don't think PhpStorm knows how to use a docker-compose binary inside WSL2.

**Comment by btantlinger** [comment]
> One year letter and after many JetBrains and WSL tickets... TADA PhpStorm 2023.2 WSL version: 1.3.14.0 PhpStorm -> Settings -> Build -> Docker -> Tools -> docker.exe binary only from [here](https://download.docker.com/win/static/stable/x86_64/) ![image](https://user-images.githubusercontent.com/1313467/258443001-4e937526-265f-4fb7-b34e-3c0cb6a87e2b.png)

@alexbogias  are you saying this is working now?  Using same versions of everything  but still having the issue here.  Do you mean the docker.exe from the link needs to be on the system path?

**Comment by alexbogias** [comment]
@btantlinger Yeah add the path you extracted it to windows system env variables and also to Settings -> Build -> Docker -> Tools -> Docker binary

**Comment by ChristianAdamski** [comment]
![grafik](https://github.com/user-attachments/assets/8052d6e8-f616-44eb-989c-04f462439986)
![grafik](https://github.com/user-attachments/assets/4cbdbbb3-afb7-4a8d-b377-55c69799350c)
![grafik](https://github.com/user-attachments/assets/c24b9145-988c-43db-91d9-71f5a7a64f92)

**Comment by ChristianAdamski** [comment]
![grafik](https://github.com/user-attachments/assets/3e9c56eb-8ea2-4691-8163-50cd47c6a556)

**Comment by ChristianAdamski** [comment]
ddev version v1.23.4

Chocolatey / The Windows side of ddev seems to not install docker anymore?

**Comment by ChristianAdamski** [comment]
Independent of DDEV, I do not understand how this is supposed to work in PHPStorm. People using WSL2 and docker within is probably the standard case by now. Where would any windows binaries come into play?

**Comment by rfay** [comment]
@ChristianAdamski Please open a new issue, probably in the ddev/ddev repository. First, get DDEV working, then work on PhpStorm. 

If you follow the instructions for installing on WSL2, you will see it try to install the Windows `ddev.exe` so that it can escalate managing /etc/hosts when that is needed. To understand it all, see https://ddev.com/blog/ddev-name-resolution-wildcards/

**Comment by ChristianAdamski** [comment]
@rfay With all due respect to your amazing work, please don't ignore qhat I posted. None of this relates to hosts or routing.

The issue is in the description of the DDEV PhpStorm Setup:

> In PhpStorm, navigate to Preferences → Build, Execution, Deployment → Docker → Tools, and set the docker-compose executable to the full path of your .ddev/bin/docker-compose file relative to your home directory.
> 
> If you’re using WSL2 and running PhpStorm on the Windows side, PhpStorm can’t use docker-compose from WSL2, so configure Docker Desktop in Settings → General to “Use Docker Compose V2” and use a recent version of Docker Desktop.

My problem is the second sentence.

**Comment by ChristianAdamski** [comment]
I did download and install Docker Desktop on the Windows Site now:

![grafik](https://github.com/user-attachments/assets/7d62b094-1600-4293-b9cf-22c5102df6e1)

I can confirm, that his generally works:

![grafik](https://github.com/user-attachments/assets/a384b487-0bfa-4019-aff4-6a2f9ee53be3)

but in PhpStorm with ddev-plugin configured PHP Interpreter, it fails to function:

![grafik](https://github.com/user-attachments/assets/0cf79bbf-6eb2-4119-a635-3ad846d9f97d)

**Comment by ChristianAdamski** [comment]
ddev itself within WSL2 works perfectly fine and I can reach the website on the Windows site as well

**Comment by alexbogias** [comment]
@ChristianAdamski Try update phpstorm to 2024.2.
Works for me
![image](https://github.com/user-attachments/assets/d29d862f-9e66-4494-a414-fb3b6bb4bd91)

![image](https://github.com/user-attachments/assets/c6e10f4d-15d3-4968-8f41-0e311c176492)

No matter the red focus ring in docker compose executable

**Comment by ChristianAdamski** [comment]
Interesting.

I am on PhpStorm 2024.2.1

Differences I note from your screenshot:

- for me the plugin does configure the the path to the compose file as a relative path (see my screenshot)
- chocolatey does not install docker anymore for me:

![grafik](https://github.com/user-attachments/assets/a9892672-3942-458f-8df8-a4c7d4db2d66)

I'm currently trying out to set up Docker and Interpreter manually.

**Comment by rfay** [comment]
It would help if you'd do a `ddev debug test` and provide a link to it as a gist or provide it as a file. Then we'd understand your setup. It does look like your PhpStorm is accidentally using a Windows-side configuration instead of a WSL configuration.

**Comment by ChristianAdamski** [comment]
I was successful by deleting existing settings and adding things manually in PhpStorm. DDEV itself was already working fine.

1.) Installed Docker Desktop on the Windows side and set paths in PHPStorm to those Windows executable
2.) Pointed PHPStorm to Docker -> WSL
3.) Added new CLI Interpreter, which resulted in the full path to the compoose file within WSL. This is probably the important part?

The rest then just worked.

**Comment by ChristianAdamski** [comment]
So, in conclusion: I do think the ddev - phpstorm plugin set things up slightly incorrectly. Also, Docker Desktop seems to be necessary until PhpStorm learns to use docker compose in WSL. There are tickets for this at intellij, but seems to be no priority for them.

So maybe close this? But add a link in the official readme as a workaround for other people running into this?

**Comment by ChristianAdamski** [comment]
Either way: thanks for everybodys help & patience. This is tough to debug for me.

**Comment by alexbogias** [comment]
@ChristianAdamski I got the binary there
![image](https://github.com/user-attachments/assets/9c19bf9c-f7b1-4457-95d9-298c7604b85d)
Did you tried to install Docker-CLI from choco?
![image](https://github.com/user-attachments/assets/eb3de62c-77a4-4baf-8966-744bcbc90284)

**Comment by rfay** [comment]
I was wondering if you had Docker Desktop installed, but were trying to use docker-ce.

**Comment by ChristianAdamski** [comment]
@rfay I never used docker on Windows knowingly. I'm 99% sure, Docker Desktop was installed by ddev/chocolatey before? It wasn't now, so I installed it manually to please PhpStorm.

Docker-CE I only know from linux.

**Comment by ChristianAdamski** [comment]
Just for the record: DDEV is such an amazon tool and this integration plugin as well. I know that at least half the problems are caused by PhpStorm itself.

So long story short: I appreaciate all the work.
